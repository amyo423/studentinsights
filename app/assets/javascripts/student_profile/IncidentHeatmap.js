import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import {toMomentFromTimestamp} from '../helpers/toMoment';
import {toSchoolYear, firstDayOfSchool} from '../helpers/schoolYear';
import {withDefaultNowContext} from '../testing/NowContainer';
import DisciplineScatterPlot, {getincidentTimeAsMinutes} from '../components/DisciplineScatterPlot';

export default class IncidentHeatmap extends React.Component {
  constructor(props) {
    super(props);
  }

  toolTipFormatter() {
    return `${this.point.type} - ${this.point.date}`;
  }

  //Discipline patterns are more likely to be meaningful within a single school year
  filterToSchoolYear(incidents) {
    const {nowFn} = this.context;
    const currentSchoolYear = toSchoolYear(nowFn());
    const cutoffDay = firstDayOfSchool(currentSchoolYear);
    return incidents.filter(incident => {
      const incidentMoment = toMomentFromTimestamp(incident.occurred_at);
      return incidentMoment.isAfter(cutoffDay);
    });
  }

  render() {
    const categories = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    const incidentsThisYear = this.filterToSchoolYear(this.props.incidents);
    const seriesData = incidentsThisYear.map(incident => {
      const x = categories.indexOf(moment.utc(incident.occurred_at).format("ddd"));
      const y = getincidentTimeAsMinutes(incident);
      const date = moment.utc(incident.occurred_at).format("MM/DD/YYYY");
      const type = incident.incident_code;
      return {x, y, date, type};
    });
    return(
      <DisciplineScatterPlot
        id={"string"}
        animation={false}
        titleText={"Incidents by Day and Time [School Year]"}
        categories={{categories: categories}}
        seriesData={seriesData}
        measureText={"Time of Incident"}
        studentChart={true}
        toolTipFormatter={this.toolTipFormatter}/>);
  }
}

IncidentHeatmap.propTypes = {
  incidents: PropTypes.array.isRequired
};
IncidentHeatmap.contextTypes = {
  nowFn: PropTypes.func.isRequired
};
