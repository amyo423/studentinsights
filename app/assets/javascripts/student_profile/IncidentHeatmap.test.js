import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import IncidentHeatmap from './IncidentHeatmap';

function testProps() {

  return { incidents: [{
    "id": 45,
    "incident_code": "ABC",
    "created_at": "2018-03-01T14:22:11.437Z",
    "updated_at": "2018-03-01T14:22:11.437Z",
    "incident_location": "Classroom",
    "incident_description": "Something happened recently.",
    "occurred_at": "2018-03-02T08:22:11.437Z",
    "has_exact_time": true,
    "student_id": 42
  }, {
    "id": 46,
    "incident_code": "XYZ",
    "created_at": "2016-02-22T14:22:11.437Z",
    "updated_at": "2016-02-22T14:22:11.437Z",
    "incident_location": "Classroom",
    "incident_description": "Something else happened a long time ago...",
    "occurred_at": "2016-02-24T08:10:11.543Z",
    "has_exact_time": true,
    "student_id": 42
  }]};
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(<IncidentHeatmap {...props} />, el);
});


it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(<IncidentHeatmap {...props} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

