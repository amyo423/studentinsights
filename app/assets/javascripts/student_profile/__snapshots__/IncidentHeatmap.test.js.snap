// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="DashboardScatterPlot"
  id="string"
  style={
    Object {
      "display": "flex",
      "flex": 1,
      "padding": 10,
      "width": "100%",
    }
  }
>
  <div
    className="HighchartsWrapper"
    style={
      Object {
        "flex": 1,
      }
    }
  />
</div>
`;
